### ISSUE TITLE
Remove use of System.Random in MonadToss

### ISSUE THREAD

**Issue by rrnewton** [rrnewton/concurrent-skiplist#4]
Oops... inside the monad-toss business there is a performance bug.  It uses a very slow RNG.

@fryguybob recommended ... which was it?  pcg-random?  Was that the one that offered good parallel scalability.

I think there was also another perf bug we discussed at ICFP, but I've forgotten it now.

**Comment by fryguybob** [comment]
Yes, pcg-random.  Specifically the [`System.Random.PCG.Fast.Pure`](https://hackage.haskell.org/package/pcg-random-0.1.3.2/docs/System-Random-PCG-Fast-Pure.html) module where `Gen` is a `MutableByteArray`.  I also changed [this line](https://github.com/cchalmers/pcg-random/blob/pure/src/System/Random/PCG/Fast/Pure.hs#L185) to bump the bytes allocated to 48 rounding the whole `MutableByteArray` allocated to one 64-byte cache-line on my 64-bit machine to avoid contention between PRNG's allocated together.  Without that I observed bad performance in the wild.

**Comment by rrnewton** [comment]
Is that one line patch linked there look correct vis a vis what you said?

**Comment by fryguybob** [comment]
That looks right to me.  `newByteArray` allocates an `StgArrWords` which has a `StgHeader` (8 bytes) then an `StgWord` (8 bytes) then the number of bytes requested (48 bytes) giving a total of `8+8+48=64` bytes.  So the 8 bytes that PCG uses will always be on a different cache-line then another PCG (modulo associativity).
